### PR TITLE
Add ember-cli support

### PR DESCRIPTION
**Pull request by igillis**
This diff turns ember-widgets into an ember-cli addon. It introduces the following breaking changes:

- View helpers have been dropped in favor of components.
- Some components have been renamed to be consistent with ember-cli style.

Tests still don't work. There is a build system in `packaging/` for generating old, globals-style builds, but second breaking change above still applies to that build. 

I'm not really sure how we should go about reviewing this. I think you can manually look at the stuff in packaging/ and for the rest you're best off just checking this branch out and running the demo app to mess around with stuff.

@ebryn @azirbel @bjcohen @cyril-sf 

Added by Alex - major questions to answer

- [x] How does the release & versioning process work? **Answer:** Same as it used to (grunt-release-it)
- [x] How will we host the demo app? (GH pages again? Do we need another build step?) **Answer:** It's in `tests/dummy`. Thanks Ember-CLI.
- [x] Should all addon code go in the addon folder?
- [x] When should we update documentation? **Answer:** Now.
- [x] Can we make the globals build totally backwards compatible? Might just be a matter of adding some special cases to `globals.js`/`registry.js` or changing some file names. **Answer:** No. At the very least, use of view helpers will be deprecated in favor of components syntax.

Ian's TODOs
- [x] Try moving to `addon/` folder (failed. Issue with coffeescript and ember addons)
- [x] Fix file copying derps (CHANGELOG, VERSION, etc.)
- [x] Add license .hbs file
- [x] Document component name changes
- [x] Add comments and a README to packaging
- [x] Remove .DS_Store (.gitignore?)
- [x] Fix lint errors
- [x] Fix fonts
- [x] Add Gruntfile + release-it task
- [x] Update github pages docs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,40 +6,38 @@ A collection of small widgets, easy to drop into place as Ember Components.
...
+```
+bower_components/ember-widgets/packaging/dist/ember-widgets.js
 ```
+in your app.
 
-To view examples, start the node server. From the root directory:
+## Running
 
-`$ node examples.js`.
+* `ember server`
```
Comment (azirbel) [inline README.md]: should add `npm install`

**Comment by azirbel** [comment]
We probably need to think about our release/versioning process. I like having a `VERSION` file and other versioning information, and while we used `grunt-release-it` for this before, we may have to find a new way now.

**Comment by azirbel** [comment]
Looks like `app/.DS_Store` was committed, that should be removed

**Comment by azirbel** [comment]
We shouldn't remove `CHANGELOG` and `CONTRIBUTING` files

**Comment by azirbel** [comment]
Speaking of `CHANGELOG`, is there a place you've described the name changes for widgets? That's pretty important for migrating users. You could use `CHANGELOG`, or copy the migration guides from http://addepar.github.io/ember-table/#/ember-table/migration-guides.

**Comment by azirbel** [comment]
I think all addon code should go in the `addon` folder so that it's in its own namespace. If the build process is smart enough, this would also allow consuming apps to include individual widgets and therefore reduce the library's footprint. In https://github.com/Addepar/ember-table-addon/tree/azirbel/demo-app-js I have all code in the `addon` folder and it seems to work for me so far.

**Comment by azirbel** [comment]
`ember server` gives coffeelint warnings that lines exceed line length; that should be easy to fix

**Comment by azirbel** [comment]
Looks like you have the same fonts problem I do. We should find a way to load Univers locally while running the demo app, because the layout/design looks a lot better with it.

**Comment by azirbel** [comment]
Of course, all documentation needs updating ;) we could do that as a separate task once this is done, but we should add a note/make a ticket or something.

**Comment by azirbel** [comment]
The license page is broken, probably you just forgot the handlebars file

**Comment by azirbel** [comment]
(I'm commenting everything here because it's hard to find everything in the diff). `transitionend` should be named `TransitionEnd`

**Comment by azirbel** [comment]
To keep track of major "question" issues, I edited your PR description to add checkboxes. That way we won't forget about them.

**Comment by azirbel** [comment]
Commented code in `packaging/Brocfile.js`. Would be great to have a lot of real comments in those files, since they are very confusing to me! But then I'm a Broccoli newbie.

**Comment by azirbel** [comment]
Er, looks like I forgot to say this on the official diff: :+1: :zap: :green_heart: :smile: :sunny: GREAT WORK! SO EXCITED TO SEE THIS HAPPENING! :dolphin: :car: :boom: :sparkles: :sunglasses:

**Comment by igillis** [comment]
Good call on the checkboxes. I added another and a TODO section for myself.

- I had some issues with the `addon/` folder originally, but I'll take another stab.
- We can still use grunt-release-it right? It seem pretty orthogonal to the actual build process.
- Which documentation are you talking about? The stuff in gh_pages? I think we have to do it now. I tried to update it as I went but there will be stuff I missed.
- I haven't thought about gh_pages versioning/deployment at all -- not really sure how it works now tbh

**Comment by igillis** [comment]
@azirbel could you take a look at the line length changes to `app/mixins/dom-helper.coffee`? I think you wrote the original and it was a bit dense -- not 100% confident in my changes.

**Comment by azirbel** [comment]
@igillis took a look, made two comments. I didn't write the code but it all looks sensible to me

File: tests/dummy/app/styles/typography.less [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 @font-face {
     font-family : "SourceCodePro-Black";
-    src: url('../fonts/sourceCodePro/SourceCodePro-Black.woff') format('woff'),
+    src: url('fonts/sourceCodePro/SourceCodePro-Black.woff') format('woff'),
```
Comment (azirbel) [inline tests/dummy/app/styles/typography.less]: this one appears to be a mistake

**Comment by azirbel** [comment]
@igillis I figured out why Univers fonts weren't loading, check out https://github.com/Addepar/ember-table-addon/commit/be4f75868e68f745348700aaeab0333db0d9a22f

**Comment by azirbel** [comment]
@igillis also figured out how to publish to gh-pages. We're lucky since `grunt` and `broccoli` can co-exist, as I still think `grunt-release-it` is the best tool for the job. (Another option: https://github.com/poetic/ember-cli-github-pages, but I like `grunt-release-it` because it can automate everything including your build command).

Check out https://github.com/Addepar/ember-table-addon/commit/424784fb96e845feda691e1f3fde6d1d6341cbb0. Nice features include being able to run `ember build --environment="production"` as the build command for `grunt-release-it`. It's necessary to have separate prod and development builds, because gh-pages will have a path like `http://addepar.github.io/ember-table-addon/simple` while your app expects `http://addepar.github.io/simple` unless configured otherwise. I was able to release ember-table-addon (I have the advantage of having a separate repository to test with) to http://addepar.github.io/ember-table-addon/simple (some pages don't quite work, looking into that issue now).

**Comment by igillis** [comment]
Awesome. Good stuff. I'll probably just copy that stuff right over.

**Comment by igillis** [comment]
Ok so I think everything is in place except moving to the `addon/` folder. 

See the discussion at https://github.com/ember-cli/ember-cli/issues/3034, but it sounds like coffeescript addons are not fully supported yet. We might be out of luck until it matures a little bit more (or, of course, we can contribute over there).

**Comment by azirbel** [comment]
Ah, that's too bad. Well in that case I think we should try to finish this conversion and keep everything in the `app` folder, so we make sure not to get stuck with two versions of the code. I'll keep reviewing today.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+---
+language: node_js
```
Comment (cyril-sf) [inline .travis.yml]: Are we using Travis?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -56,16 +56,19 @@ provide as much detail and context as possible.
...
-dependencies) locally installed.
+compiled using [ember-cli](http://www.ember-cli.com/).
 
-The `src` folder contains all the source code for the ember-widgets project
-itself. The `app` folder contains the sample app and documentation, and should
-be updated whenever changes will add functionality, modify or remove existing
-functionality, or change public APIs. The `dist` folder contains the compiled
-ember-widgets code and should not be directly modified but instead generated
-using `grunt dist`.
+The `app` folder contains all the source code for the ember-widgets project
```
Comment (cyril-sf) [inline CONTRIBUTING.md]: I thought that the source of an addon should live in `addon` not `app`?

File: app/components/popover-link-component.coffee [hunk 5]
```diff
@@ -10,14 +13,6 @@ Ember.Widgets.PopoverLinkComponent = Ember.Component.extend
   rootElement: '.ember-application'
   fade: yes
 
-  willDestroy: ->
```
Comment (cyril-sf) [inline app/components/popover-link-component.coffee]: Why did you remove this?

File: app/initializers/ember-widgets-config.coffee [hunk 6]
```diff
@@ -0,0 +1,14 @@
+`import Ember from 'ember'`
+
+EmberWidgetsConfigInitializer = Ember.Object.create
```
Comment (cyril-sf) [inline app/initializers/ember-widgets-config.coffee]: Do you need to create an object here?

File: app/mixins/dom-helper.coffee [hunk 7]
```diff
@@ -11,19 +13,20 @@ Ember.Widgets.DomHelper = Ember.Mixin.create
...
 
-  # Set the selected range to the given element, with the option to collapse the selection to the
-  # beginning or end
+  # Set the selected range to the given element, with the option to collapse
+  # the selection to the beginning or end
   #   element       the element to select
   #   collapseMode  "none", "beginning", "end"
   selectElement: (document, element, collapseMode="end") ->
-    # http://stackoverflow.com/questions/1125292/how-to-move-cursor-to-end-of-contenteditable-entity
+    # http://bit.ly/1tOBYVf
```
Comment (cyril-sf) [inline app/mixins/dom-helper.coffee]: I'd rather keep the original URL

File: app/components/popover-link-component.coffee (hunk 5 above)
Comment (cyril-sf) [inline app/components/popover-link-component.coffee]: ( oh I guess you would need to be rebase, which is going to be painful :cry: )

**Comment by cyril-sf** [comment]
Pretty excited by that change!

Moving forward, I think we should get the demo app in a separate repo and have real test coverage.

**Comment by azirbel** [comment]
@igillis were you able to use this from an ember-cli app? I tried it this weekend but couldn't get it to install. Didn't have much time to look into it.

File: app/components/popover-link-component.coffee (hunk 5 above)
Comment (igillis) [inline app/components/popover-link-component.coffee]: yeah, my 'rebase' method right now is basically to to `--ours` and then manually pull in the changes

**Comment by igillis** [comment]
@azirbel I'm able to install correctly using this command in a new ember-cli app:

`ember install:addon git+ssh://git@github.com:Addepar/ember-widgets.git#igillis/ember-cli`

However, I'm running into issues at that point. I believe what should happen is that `dist/{app-name}.js` has a bunch of `defines` for all of the addon's stuff. Instead, I'm only seeing defines for the ember-widgets templates...I'm a bit worried that it's a bug in ember-cli (https://github.com/ember-cli/ember-cli/issues/2354) and not something on our end

File: .editorconfig [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# EditorConfig helps developers define and maintain consistent
+# coding styles between different editors and IDEs
+# editorconfig.org
```
Comment (azirbel) [inline .editorconfig]: :+1: really happy about this

File: .editorconfig [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+root = true
+
```
Comment (azirbel) [inline .editorconfig]: extra newline

File: .travis.yml (hunk 3 above)
Comment (azirbel) [inline .travis.yml]: @igillis I assume this is an ember-cli default for working with Travis? Can we take it out for now, but hopefully someday add it back ... and get Travis ... ?

File: Gruntfile.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
+module.exports = function(grunt) {
```
Comment (azirbel) [inline Gruntfile.js]: add `'use strict';`

File: Gruntfile.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  grunt.initConfig({
+    'release-it': {
+      options: {
+        'pkgFiles': ['package.json', 'bower.json'],
+        'commitMessage': 'Release %s',
+        'tagName': 'v%s',
+        'tagAnnotation': 'Release %s',
+        'increment': 'patch',
+        'buildCommand': 'ember build --environment="production"',
+        'distRepo': '-b gh-pages git@github.com:addepar/ember-widgets-addon',
```
Comment (azirbel) [inline Gruntfile.js]: modify to `ember-widgets`

File: LICENSE.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+The MIT License (MIT)
```
Comment (azirbel) [inline LICENSE.md]: I think we use BSD.

We should make sure this is standardized, but at the least, be careful not to add new licenses

File: README.md [hunk 13]
```diff
@@ -6,40 +6,32 @@ A collection of small widgets, easy to drop into place as Ember Components.
...
 ### JS Bin Starter Kit
 
 http://emberjs.jsbin.com/hebexa/1/edit
 
-### Installation
-
-With bower: `bower install ember-widgets --save`
-
-Or, just include `dist/js/ember-widgets.js` and `dist/css/ember-widgets.css` in your app.
+## Ember-CLI Installation
```
Comment (azirbel) [inline README.md]: I think this section should be `npm install --save-dev ember-widgets`, as we should put "how to use" instructions at the top of the app. The `git clone...` instructions should be father down in the README, in a "devloping" section.

File: app/components/modal-component.coffee [hunk 14]
```diff
@@ -39,21 +47,21 @@ Ember.Component.extend Ember.Widgets.StyleBindingsMixin,
     templateName: 'modal-footer'
 
   _headerViewClass: Ember.computed ->
-    headerViewClass = @get 'headerViewClass'
+    headerViewClass = Ember.get @, 'headerViewClass'
```
Comment (azirbel) [inline app/components/modal-component.coffee]: use `this` instead of `@` when used on its own. Applies to a few places below as well.

Comment (azirbel) [inline app/components/modal-component.coffee]: actually, why did you make this change? The old way looks better...

File: app/components/select-component.coffee [hunk 15]
```diff
@@ -164,12 +107,12 @@ Ember.AddeparMixins.ResizeHandlerMixin,
     Ember.run this, @updateDropdownLayout
 
   # TODO(Peter): consider calling this optionViewClass?
-  itemView: Ember.computed ->
+  itemView: Ember.computed(->
```
Comment (azirbel) [inline app/components/select-component.coffee]: also unclear why you're changing these

File: app/mixins/dom-helper.coffee (hunk 7 above)
Comment (azirbel) [inline app/mixins/dom-helper.coffee]: Agree

File: packaging/globals.js [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    'mixins',
+    'controllers'
+  ];
+};
+
+Globals.prototype = Object.create( Writer.prototype );
+Globals.prototype.constructor = Globals;
+
+Globals.prototype.write = function (readTree, destDir) {
+  var self = this;
```
Comment (azirbel) [inline packaging/globals.js]: I take it this is because we've copied the globals packaing code from elsewhere - usually we should use `_this`. Add a comment!

File: packaging/globals.js [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+          modules.push("'" + module + "'");
+          dependencies.push('__dependency' + (dependencies.length+1) + '__');
+
+           /*
+            * Component, View, etc
+            * Take the folder name, capitalize the first letter, and remove the
+            * pluralization
+            */
+          var objectType = self.capitalize(parts[0]).replace(/s$/, '');
```
Comment (azirbel) [inline packaging/globals.js]: ick. This should use the ember pluralizer instead. Again, adding a comment is ok.

File: packaging/globals.js [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+          /*
+           * Converts from foo-bar-baz to FooBarBaz
+           *   Assumes file name is dasherized and lower case
+           *   Take the file name, strip the extension, tokenize by the hyphen,
+           *   for each token capitalize the first character and join.
+           */
+          var objectName = parts[1]
```
Comment (azirbel) [inline packaging/globals.js]: pretty sure there are functions that do this, too.

File: tests/dummy/app/mixins/code-pretty-print.coffee [hunk 19]
```diff
@@ -0,0 +1,8 @@
+`import Ember from 'ember'`
+
+CodePrettyPrintMixin = Ember.Mixin.create
+  didInsertElement: ->
+    @_super()
+    Ember.run.next this, -> prettyPrint()
```
Comment (azirbel) [inline tests/dummy/app/mixins/code-pretty-print.coffee]: this works...? where does `prettyPrint()` come from?

File: tests/dummy/app/templates/ember-widgets/migration-guides.hbs [hunk 20]
```diff
@@ -0,0 +1,45 @@
+{{!-- To add shadow main-content-container-shadow --}}
+<div class="col-md-10 col-md-offset-2 left-border main-content-container">
+  <h1>Migration Guides</h1>
+  <div class="row">
+    <div class="col-md-9">
+      <h2>Migrating to version 0.3.0 </h2>
```
Comment (azirbel) [inline tests/dummy/app/templates/ember-widgets/migration-guides.hbs]: woooooot :+1: migration guides! wow, great work.

Comment (azirbel) [inline tests/dummy/app/templates/ember-widgets/migration-guides.hbs]: @Dylnuge pressure's on, widgets and table both have migration guides now. haha

File: tests/dummy/config/environment.js [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+    // keep test console output quieter
+    ENV.APP.LOG_ACTIVE_GENERATION = false;
+    ENV.APP.LOG_VIEW_LOOKUPS = false;
+
+    ENV.APP.rootElement = '#ember-testing';
+  }
+
+  if (environment === 'production') {
+
```
Comment (azirbel) [inline tests/dummy/config/environment.js]: probably you will need to add `ENV.baseURL = '/ember-widgets/';` to make it work witih gh-pages

**Comment by azirbel** [comment]
Did another pass over the code. LGTM. I think we're pretty close to being able to merge this - I've made a checklist below of the order of events as I see them:

- [ ] Merge in and release versions for Ember 1.8, 1.9, and 1.10. Require Ember 1.10 and Handlebars 2.0 in the main project.
- [ ] Remove text editor
- [ ] Try new ember-cli that should allow putting code in the `addon` folder. I think it's worth waiting for this to prevent further breaking changes if we were to put the code temporarily into `app`.
- [ ] Address comments on this PR

At that point, we're ready to merge. We should then:

- [ ] Make sure we've pulled in any new changes to ember-widgets
- [ ] Verify that we can run the demo app
- [ ] Fork this branch to your personal repo, change the release-it config to point to your repo as well (since it has a hardcoded repo name), and try releasing. Verify that everything shows up on gh-pages. This sometimes takes a bit of fiddling with configurations.
- [ ] Make a new Ember CLI project and try importing
- [ ] Make a new old-style project using globals and try importing
- [ ] Merge

I'm happy to help with the things in the second list, and it would be cool if you mentioned when we get there, but I'm also confident that you can take care of all these things. So if it will get this PR across the finish line, feel free to do them all in quick succession and merge.

**Comment by blimmer** [comment]
It's really exciting to see this work getting done. Looks like there hasn't been activity here in about a month. Any updates on this? Right now it's looking like we are going to have to move the components we use over to an addon that I maintain for the time being.

**Comment by igillis** [comment]
@blimmer unfortunately this keeps getting pushed down the stack. 

If you can't make the current, globals-based distribution work with your project, then I think your best bet is to do something similar to https://github.com/Addepar/ember-table-addon where we just create an ember-cli app that exports the globals build.

**Comment by igillis** [comment]
Tried the most recent versions of ember-cli/ember-cli-coffeescript and it's still the case that coffeescript code in the `addon/` folder doesn't get compiled. JS code there works fine, as does CS code in `app/`.

Going to continue with everything in `app/` unless someone has a solution.

File: tests/dummy/app/mixins/code-pretty-print.coffee (hunk 19 above)
Comment (igillis) [inline tests/dummy/app/mixins/code-pretty-print.coffee]: I'm assuming from whatever class it gets mixed into? This is just copied code.

Comment (azirbel) [inline tests/dummy/app/mixins/code-pretty-print.coffee]: I think when I made this comment I didn't know what prettyPrint was. It's probably defined globally by the pretty-print dependency ... looks fine

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -148,10 +156,10 @@ included in the project:
 license your work under the terms of the [BSD License](LICENSE).
 
 
-## Running tests
+## Running tests (WIP -- do not use!)
```
Comment (cyril-sf) [inline CONTRIBUTING.md]: What is the problem?

**Comment by cyril-sf** [comment]
@igillis: let's circle back with @azirbel to see how we can bring that PR to the finish line

**Comment by mrloop** [comment]
Not sure what the current status is with this branch. In order to use multi-select in a project using EmberJS 1.11 and ember-cli 0.2.3 I needed to make the following changes https://github.com/mrloop/ember-widgets/commit/c94184c822679588c9031dba9ffb14af12bd8dcd

**Comment by embooglement** [comment]
@igillis @cyril-sf What exactly is the status of this branch?

**Comment by igillis** [comment]
Last I recall, we were at a point where everything was using modules and we had a backwards compatible globals build, but I was running into issues getting ember-cli to play nicely with coffeescript. Stuff in `addon` wasn't getting compiled, only stuff in `app`. We also still needed to do some more testing.

I think we should adopt table's approach and convert to JS first.

**Comment by blimmer** [comment]
A few pieces of advice:

* I think ember-cli-coffeescript should work with addons - there's other projects doing this, so I'd make sure you're on the most current. Files in /app get merged into the consuming project, so it might freak out if the consuming app doesn't use coffeescript. So, at a minimum, you might need to convert app/ files to JS

* It's a good idea to convert to ES210* style code - we did it on our project, it takes forever but is very worth it. There's a good [tool](https://github.com/eventualbuddha/decaffeinate) that can help convert to ES201* flavor JS code
